### PR TITLE
Issue 5008 fixed

### PR DESCRIPTION
**Pull request by notoraptor**
Hello ! This is my proposal for the issue 5008 (NumPy C alternative for [sd]gemm_).

### CODE REVIEW

**Comment by slefrancois** [comment]
jenkins test this please

**Comment by slefrancois** [comment]
Jenkins found a flake8 error, see the error message to correct it.

**Comment by notoraptor** [comment]
I have corrected the flake8 errors !

**Comment by slefrancois** [comment]
add to whitelist

File: theano/tensor/alt_sgemm.c [hunk 1]
```diff
@@ -0,0 +1,77 @@
+/**
+C Implementation of sgemm_ based on NumPy
+Used instead of blas when Theano config flag blas.ldflags is empty.
+**/
+inline PyObject* alt_transpose(PyArrayObject* o) {
```
Comment (nouiz) [inline theano/tensor/alt_sgemm.c]: This function is used only once and is very short. Remove it and inline the code manually.

idem for alt_matrix_matrix_product and alt_matrix_matrix_product2

File: theano/tensor/alt_sgemm.c [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ * NB: We can more optimize this function (for example, when alpha == 0).
+ * */
+void sgemm_(char* TRANSA, char* TRANSB, 
+			const int* M, const int* N, const int* K,
+			const float* ALPHA,  float* A, const int* LDA, 
+			 float* B, const int* LDB, const float* BETA, 
+			float* C, const int* LDC) {
+	if(*M < 0 || *N < 0 || *K < 0 || *LDA < 0 || *LDB < 0 || *LDC < 0)
+		return;
+	if(C == NULL)
```
Comment (nouiz) [inline theano/tensor/alt_sgemm.c]: why this can happen? See also the comment on line 28.

File: theano/tensor/alt_sgemm.c [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+}
+inline PyObject* alt_op(char* trans, PyArrayObject* matrix) {
+	return (*trans == 'N' || *trans == 'n') ? (PyObject*)matrix : alt_transpose(matrix);
+}
+/* sgemm
+ * We assume that none of these 13 pointers passed as arguments is null.
+ * NB: We can more optimize this function (for example, when alpha == 0).
+ * */
+void sgemm_(char* TRANSA, char* TRANSB, 
+			const int* M, const int* N, const int* K,
```
Comment (nouiz) [inline theano/tensor/alt_sgemm.c]: Don't use tab. Use 4 spaces indentation instead. We try to get rid of tab everywhere in Theano.

File: theano/tensor/alt_sgemm.c [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+}
+inline PyObject* alt_matrix_matrix_product2(PyObject* o1, PyObject* o2, PyObject* out) {
+	return PyArray_MatrixProduct2(o1, o2, (PyArrayObject*)out);
+}
+void alt_scalar_matrix_product_in_place(float scalar, float* matrix, int size_to_compute) {
+	for(int i = 0; i < size_to_compute; ++i) {
+		matrix[i] *= scalar;
+	}
+}
+void alt_matrix_sum_in_place(float* A, float* B, float* out, int size_to_compute) {
```
Comment (nouiz) [inline theano/tensor/alt_sgemm.c]: This isn't inplace. Update the name.

File: theano/tensor/alt_sgemm.c [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		kb = *N;
+	else
+		kb = *K;
+	npy_intp dims_A[2] = {*LDA, ka};
+	npy_intp dims_B[2] = {*LDB, kb};
+	PyObject* matrix_A = PyArray_SimpleNewFromData(2, dims_A, NPY_FLOAT32, A);
+	PyObject* matrix_B = PyArray_SimpleNewFromData(2, dims_B, NPY_FLOAT32, B);
+	PyObject* op_A = alt_op(TRANSA, (PyArrayObject*)matrix_A);
+	PyObject* op_B = alt_op(TRANSB, (PyArrayObject*)matrix_B);
+	if(*BETA == 0) {
```
Comment (nouiz) [inline theano/tensor/alt_sgemm.c]: C is not necesserily contiguous, so call on lines 64, 69 and 79 could return wrong answer. Pass M, N and LDC  in alt_scalar_matrix_product_in_place and alt_matrix_sum_in_place to treat them as matrix.

**Comment by notoraptor** [comment]
Pull request updated.

**Comment by notoraptor** [comment]
Updated.

**Comment by notoraptor** [comment]
EDIT: I made some mistake in the previous commit message:

_"
All tests succeed (with blas.ldflags empty) for:
* test_abstract_conv.py in theano/tensor/nnet/tests/
"_

All test succeed for test_abstract_conv.py**:TestCorrConv2d** only, I have not executed the other tests in this file.

Sorry !

**Comment by notoraptor** [comment]
New update !

Code for [sd]gemm_ radically changed, and many (if not all) gemm-related tests passed.

**Comment by lamblin** [comment]
Thanks for the update, I should be able to review tomorrow.

**Comment by nouiz** [comment]
jenkins test this.

**Comment by nouiz** [comment]
I merged a PR that now make this PR have conflict with the master. Can you rebase it?

**Comment by notoraptor** [comment]
Updated !

**Comment by notoraptor** [comment]
New update !

**Comment by notoraptor** [comment]
All tests passed.

```
TEST: theano/tensor/tests/test_blas_c.py
...........S.S.SSSSSSSSSSSSSSSSSSSSSSSS...SS
----------------------------------------------------------------------
Ran 44 tests in 16.226s

OK (SKIP=28)
TEST: theano/tensor/tests/test_blas_scipy.py
..............
----------------------------------------------------------------------
Ran 14 tests in 18.115s

OK
TEST: theano/tensor/nnet/tests/test_abstract_conv.py:TestBilinearUpsampling
.....
----------------------------------------------------------------------
Ran 5 tests in 33.907s

OK
TEST: theano/tensor/nnet/tests/test_abstract_conv.py:TestCpuConv3d
.................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSS...................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSS..
----------------------------------------------------------------------
Ran 98 tests in 43.732s

OK (SKIP=60)
TEST: theano/tensor/tests/test_blas.py
..............................................................................................................
----------------------------------------------------------------------
Ran 110 tests in 82.597s

OK
TEST: theano/tensor/nnet/tests/test_abstract_conv.py:TestAbstractConvNoOptim
....................
----------------------------------------------------------------------
Ran 20 tests in 98.397s

OK
TEST: theano/tensor/nnet/tests/test_abstract_conv.py:TestCpuConv2d
.......................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS..
----------------------------------------------------------------------
Ran 402 tests in 148.494s

OK (SKIP=252)
TEST: theano/tensor/nnet/tests/test_corr.py
Tests that basic correlations work for odd and even ... ok
Checks dtype upcast for CorrMM methods. ... ok
Tests correlation where filter dilation != (1,1) ... ok
Tests basic correlation in full mode and case where filter ... ok
test_img_kernel_same_shape (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
test_infer_shape_forward (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
test_infer_shape_gradI (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
test_infer_shape_gradW (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
Tests scenario where filter_shape[1] != input_shape[1] ... ok
test_non_contiguous (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
Tests correlation where the {image,filter}_shape is a Constant tensor. ... ok
Tests correlation where subsampling != (1,1) ... ok
Make sure errors are raised when image and kernel are not 4D tensors ... ok

----------------------------------------------------------------------
Ran 13 tests in 179.799s

OK
TEST: theano/tensor/nnet/tests/test_abstract_conv.py:TestCorrConv3d
..................................................................................................
----------------------------------------------------------------------
Ran 98 tests in 315.026s

OK
TEST: theano/tensor/nnet/tests/test_abstract_conv.py:TestCorrConv2d
..................................................................................................................................................................................................................................................................................................................................................................................................................
----------------------------------------------------------------------
Ran 402 tests in 631.067s

OK
TEST: theano/tensor/nnet/tests/test_corr3d.py
Tests that basic correlations work for odd and even ... ok
Checks dtype upcast for Corr3dMM methods. ... ok
Tests correlation where filter dilation != (1,1,1) ... ok
Tests basic correlation in full mode and case where filter ... ok
test_img_kernel_same_shape (theano.tensor.nnet.tests.test_corr3d.TestCorr3D) ... ok
test_infer_shape_forward (theano.tensor.nnet.tests.test_corr3d.TestCorr3D) ... ok
test_infer_shape_gradI (theano.tensor.nnet.tests.test_corr3d.TestCorr3D) ... ok
test_infer_shape_gradW (theano.tensor.nnet.tests.test_corr3d.TestCorr3D) ... ok
Tests scenario where filter_shape[1] != input_shape[1] ... ok
test_non_contiguous (theano.tensor.nnet.tests.test_corr3d.TestCorr3D) ... ok
Tests correlation where the {image,filter}_shape is a Constant tensor. ... ok
Tests correlation where subsampling != (1,1,1) ... ok
Make sure errors are raised when image and kernel are not 5D tensors ... ok

----------------------------------------------------------------------
Ran 13 tests in 291.664s

OK
```

File: theano/tensor/alt_gemm_template.c [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    char* TRANSA, char* TRANSB, 
+    const int* M, const int* N, const int* K,
+    const %(float_type)s* ALPHA, %(float_type)s* A, const int* LDA, 
+    %(float_type)s* B, const int* LDB, const %(float_type)s* BETA, 
+    %(float_type)s* C, const int* LDC) {
+    /* NB: it seems that matrix+matrix and scalar*matrix functions
+     * defined above do not allocate iterator for a matrix with 0
+     * content, that is a matrix whose nrow*ncol == 0. As these
+     * functions actually work with M*N matrices (op(A)*op(B) and/or C),
+     * I think that we could just return if M or N is null. */
```
Comment (lamblin) [inline theano/tensor/alt_gemm_template.c]: If M or N is 0, indeed we can exit.

File: theano/tensor/alt_gemm_template.c [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    if(*M < 1 || *N < 1 || *K < 0 || *LDA < 0 || *LDB < 0 || *LDC < 0)
```
Comment (lamblin) [inline theano/tensor/alt_gemm_template.c]: I think that if any of these is < 0, we should exit with an error (`alt_fatal_error`).

File: theano/tensor/blas.py [hunk 8]
```diff
@@ -1037,10 +1037,7 @@ def c_code(self, node, name, inp, out, sub):
         if node.inputs[0].type.dtype.startswith('complex'):
             raise utils.MethodNotDefined('%s.c_code'
                                          % self.__class__.__name__)
-        if not config.blas.ldflags:
-            return super(Gemm, self).c_code(node, name,
-                                            (_z, _a, _x, _y, _b), (_zout, ),
-                                            sub)
+        # if not config.blas.ldflags: # return super(Gemm, self).c_code(node, name, (_z, _a, _x, _y, _b), (_zout, ), sub)
```
Comment (lamblin) [inline theano/tensor/blas.py]: You can remove the commented-out code here and on the following lines.

File: theano/tensor/alt_gemm_template.c (hunk 6 above)
Comment (notoraptor) [inline theano/tensor/alt_gemm_template.c]: OK !

File: theano/tensor/alt_gemm_template.c [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        while(count) {
+            *((%(float_type)s*)data) *= scalar;
+            data += stride;
+            --count;
+        }
+    } while(get_next(iterator));
+    NpyIter_Deallocate(iterator);
+}
+/*Matrix+Matrix function. Compute (coeffA * matrixA) + (coeffB * matrixB)
+ * Remark: This function actually sums a C-contiguous matrix (alpha*op(A)*op(B)) with a F-contiguous matrix (beta*C)
```
Comment (notoraptor) [inline theano/tensor/alt_gemm_template.c]: Do you talk about the comment line ? If yes, I can remove it, but it is still true (that this function sums a C-contiguous with a F-contiguous matrix; I have made this comment to signal that there is a performance limitation here, as the iteration can't be as quickly as possible) !

File: theano/tensor/alt_gemm_template.c [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    NpyIter_IterNextFunc* get_next = NpyIter_GetIterNext(iterators, NULL);
+    char** data_ptr_array = NpyIter_GetDataPtrArray(iterators);
+    do {
+        %(float_type)s* from_A = (%(float_type)s*)data_ptr_array[0];
+        %(float_type)s* from_B = (%(float_type)s*)data_ptr_array[1];
+        *from_B = (*ALPHA)*(*from_A) + (*BETA)*(*from_B);
+    } while(get_next(iterators));
+    NpyIter_Deallocate(iterators);
+}
+PyObject* alt_op_without_copy_%(float_type)s(int transposable, %(float_type)s* M, int nrow, int ncol, int LDM) {
```
Comment (notoraptor) [inline theano/tensor/alt_gemm_template.c]: Done ("transposable" renamed to "to_transpose", as it indicates an operation to do) !

File: theano/tensor/alt_gemm_template.c (hunk 7 above)
Comment (notoraptor) [inline theano/tensor/alt_gemm_template.c]: Corrected !

File: theano/tensor/blas.py (hunk 8 above)
Comment (notoraptor) [inline theano/tensor/blas.py]: Done (also done in other .py files related to empty ldflags).

File: theano/tensor/alt_gemm_template.c [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    // By default, M is considered as a nrow*ncol F-contiguous matrix with LDM as stride indicator for the columns.
+    npy_intp dims[2];
+    npy_intp strides[2];
+    int flags;
+    if(transposable) {
+        dims[0] = ncol;
+        dims[1] = nrow;
+        strides[0] = dims[1] * %(float_size)d;
+        strides[1] = %(float_size)d;
+        flags = NPY_ARRAY_C_CONTIGUOUS;
```
Comment (notoraptor) [inline theano/tensor/alt_gemm_template.c]: The flag is passed to PyArray_New() function, for which it seems it just indicates the format (here C-contiguous or F-contiguous) of the data pointer, but it does not indicate to convert the data to have contiguous memory.

https://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_New

So no copy is done by PyArray_New() when we pass this flag to it. Instead, when we pass wrong flag, the returned PyArrayObject is unable to iterate correctly over his data, that leads to calculation errors later !

Or have I misunderstood the point ?

File: theano/tensor/alt_gemm_template.c (hunk 9 above)
Comment (lamblin) [inline theano/tensor/alt_gemm_template.c]: I think I was confused by the fact that you were mentioning "C", which is not a parameter of this function, but of `gemm`. I think we should not refer to the parameters of gemm, but only of this function in the comment, and reference the actual name of the parameters (for instance, "ALPHA" and "A" and not "coeffA" and "matrixA").
We should also say that it actually does that computation inplace in B, something like `B = ALPHA * A + BETA * B`.

Regarding the memory layout of the matrices, does this function actually need A to be C-contiguous and B to be F-contiguous, or even C order vs F order? Or should it work either way? Because if it is a limitation of this function, we should add some checks at the beginning of the function. If not, then we can reformulate the comment to say that it would be a performance limitation if called with matrices with different layouts.

Also, if you do the same trick with `PyArray_MatrixProduct` that you did with `PyArray_MatrixProduct2` (changing the transposed status of A and B and swapping them), then the transpose of the resulting array would be in F order, like C.

File: theano/tensor/nnet/tests/test_corr.py [hunk 12]
```diff
@@ -131,7 +130,8 @@ def sym_CorrMM(input, filters):
                                 icol:icol + dil_fil_shape2d[1]:filter_dilation[1]] * filter2d[::-1, ::-1]
                             ).sum()
 
-        utt.assert_allclose(theano_output, ref_output)
+        # utt.assert_allclose(theano_output, ref_output)
```
Comment (lamblin) [inline theano/tensor/nnet/tests/test_corr.py]: You can remove that commented-out line as well

File: theano/tensor/alt_gemm_template.c (hunk 11 above)
Comment (lamblin) [inline theano/tensor/alt_gemm_template.c]: OK, I think I misunderstood the doc for `PyArray_New()`, indeed there will not be a copy.
However, I think the `CONTIGUOUS` flags can be too string (the array is not considered C-contiguous if `LDM != dims[1]`, or F-contiguous if `LDM != dims[0]`, even though it is good enough for BLAS).
I think we may need to manually test corner cases.

File: theano/tensor/alt_gemm_template.c (hunk 9 above)
Comment (notoraptor) [inline theano/tensor/alt_gemm_template.c]: OK ! Sorry for the confusion, I will rewrite the comment and this function's parameters names so that there will be no more confusion with the gemm parameters.

---
I just realized that my point about F-contiguous vs C-contiguous sum seems useless !

In fact, the problem come from gemm implementation when BETA != 0: PyArray_MatrixProduct() returns a C-contiguous matrix (that is actually op(A)*op(B)), while C matrix must be read as F-contiguous. So the sum function actually works with a C-contiguous matrix (his first parameter) and a F-contiguous matrix (second parameter), so that the iteration with NpyIter could be fast for one matrix but less for the other.

I initially thought that the iteration would be faster if both matrices have the same "contiguousity" because we could then iterate it simultaneously just through memory order, regardless of the logic order, making the iteration more faster, and that's why I had made this comment.

But in fact, it also depends on the strides of the matrices: if both matrices have not the same strides (for example, if LDC != M), we will still have to iterate in the "logic" way, and then there will not be real performance gain.

So I think I will remove this comment !

**Comment by notoraptor** [comment]
New update. All tests passed.

```
TEST01: theano/tensor/tests/test_blas_c.py
EXEC01: theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/tests/test_blas_c.py
...........S.S.SSSSSSSSSSSSSSSSSSSSSSSS...SS
----------------------------------------------------------------------
Ran 44 tests in 15.098s

OK (SKIP=28)
TEST02: theano/tensor/tests/test_blas_scipy.py
EXEC02: theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/tests/test_blas_scipy.py
..............
----------------------------------------------------------------------
Ran 14 tests in 17.230s

OK
TEST03: theano/tensor/nnet/tests/test_abstract_conv.py:TestBilinearUpsampling
EXEC03: theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestBilinearUpsampling
.....
----------------------------------------------------------------------
Ran 5 tests in 28.873s

OK
TEST04: theano/tensor/nnet/tests/test_abstract_conv.py:TestCpuConv3d
EXEC04: theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestCpuConv3d
.................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSS...................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSS..
----------------------------------------------------------------------
Ran 98 tests in 39.859s

OK (SKIP=60)
TEST05: theano/tensor/tests/test_blas.py
EXEC05: theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/tests/test_blas.py
..............................................................................................................
----------------------------------------------------------------------
Ran 110 tests in 74.695s

OK
TEST06: theano/tensor/nnet/tests/test_abstract_conv.py:TestAbstractConvNoOptim
EXEC06: theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestAbstractConvNoOptim
....................
----------------------------------------------------------------------
Ran 20 tests in 93.155s

OK
TEST07: theano/tensor/nnet/tests/test_abstract_conv.py:TestCpuConv2d
EXEC07: theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestCpuConv2d
.......................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS..
----------------------------------------------------------------------
Ran 402 tests in 138.721s

OK (SKIP=252)
TEST08: theano/tensor/nnet/tests/test_corr.py
EXEC08: theano-cache purge && THEANO_FLAGS=optdb.max_use_ratio=7,blas.ldflags= nosetests theano/tensor/nnet/tests/test_corr.py
.............
----------------------------------------------------------------------
Ran 13 tests in 160.332s

OK
TEST09: theano/tensor/nnet/tests/test_abstract_conv.py:TestCorrConv3d
EXEC09: theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestCorrConv3d
..................................................................................................
----------------------------------------------------------------------
Ran 98 tests in 288.426s

OK
TEST10: theano/tensor/nnet/tests/test_abstract_conv.py:TestCorrConv2d
EXEC10: theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestCorrConv2d
..................................................................................................................................................................................................................................................................................................................................................................................................................
----------------------------------------------------------------------
Ran 402 tests in 595.415s

OK
TEST11: theano/tensor/nnet/tests/test_corr3d.py
EXEC11: theano-cache purge && THEANO_FLAGS=optdb.max_use_ratio=10,blas.ldflags= nosetests theano/tensor/nnet/tests/test_corr3d.py
.............
----------------------------------------------------------------------
Ran 13 tests in 265.825s

OK
```

File: theano/tensor/alt_gemm_template.c [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        PyObject* matrix_C = PyArray_New(&PyArray_Type, 2, dims_C, %(npy_float)s, strides_C, C, 0, NPY_ARRAY_C_CONTIGUOUS | NPY_ARRAY_WRITEABLE, NULL);
+        PyObject* op_A_transposed = alt_op_without_copy_%(float_type)s(!is_A_transposable, A, nrowa, ncola, *LDA);
+        PyObject* op_B_transposed = alt_op_without_copy_%(float_type)s(!is_B_transposable, B, nrowb, ncolb, *LDB);
+        PyArray_MatrixProduct2(op_B_transposed, op_A_transposed, (PyArrayObject*)matrix_C);
+        if(*ALPHA != 1.0)
+            alt_numpy_scalar_matrix_product_in_place_%(float_type)s(*ALPHA, (PyArrayObject*)matrix_C);
+        Py_XDECREF(op_B_transposed);
+        Py_XDECREF(op_A_transposed);
+        Py_XDECREF(matrix_C);
+    } else {
```
Comment (notoraptor) [inline theano/tensor/alt_gemm_template.c]: Done !

**Comment by lamblin** [comment]
I think I have an example, if slightly convoluted, that fails:
```python
import numpy as np
import theano

aval = np.ones((6, 2))
bval = np.ones((2, 7))
cval = np.arange(7) + np.arange(0, .6, .1)[:, np.newaxis]

a = theano.shared(aval[:3], borrow=True)
b = theano.shared(bval[:, :5], borrow=True)
c = theano.shared(cval[:3, :5], borrow=True)

s = theano.tensor.scalar()
upd_c = s * c + theano.tensor.dot(a, b)
f = theano.function([s], [], updates={c: upd_c})

f(0)
```
Which gives:
`ValueError: output array is not acceptable (must have the right type, nr dimensions, and be a C-Array)`

**Comment by notoraptor** [comment]
@lamblin Indeed you're right ! I am updating the code.

**Comment by notoraptor** [comment]
New update !

Special test succeeds nows, and all other tests also:

```
import numpy as np
import theano

aval = np.ones((6, 2))
bval = np.ones((2, 7))
cval = np.arange(7) + np.arange(0, .6, .1)[:, np.newaxis]

a = theano.shared(aval[:3], borrow=True)
b = theano.shared(bval[:, :5], borrow=True)
c = theano.shared(cval[:3, :5], borrow=True)

s = theano.tensor.scalar()
upd_c = s * c + theano.tensor.dot(a, b)
f = theano.function([s], [], updates={c: upd_c})

f(0)
print c.get_value()
```

Execution:

`THEANO_FLAGS=blas.ldflags= python test-numpy-gemm-fail.py`

Output:

```
[[ 2.  2.  2.  2.  2.]
 [ 2.  2.  2.  2.  2.]
 [ 2.  2.  2.  2.  2.]]
```

All "NPY_ARRAY_(F|C)_CONTIGUOUS" have been removed from the code, as NumPy seems just to ignore these flags when we pass explcit strides.

When C is not contiguous (as in this special test; C not-contiguous is evaluated with LDC != M), a new memory data is now allowed, and then copied later to C.

File: theano/tensor/alt_gemm_template.c [hunk 14]
```diff
@@ -133,14 +146,20 @@ void %(name)s(
     PyArray_MatrixProduct2(opB_transposed, opA_transposed, (PyArrayObject*)opB_trans_dot_opA_trans);
     if(*BETA == 0) {
         if(*ALPHA != 1.0)
-            alt_numpy_scale_matrix_inplace_%(float_type)s(*ALPHA, (PyArrayObject*)opB_trans_dot_opA_trans);
+            alt_numpy_scale_matrix_inplace_%(float_type)s(ALPHA, (PyArrayObject*)opB_trans_dot_opA_trans);
+        if(*LDC != *M) {
+            /* A buffer has been created to compute ALPHA*op(A)*op(B),
+             * so we must copy it to the real output, that is C. */
+            PyObject* matrix_C = alt_wrap_fortran_writeable_matrix_%(float_type)s(C, M, N, LDC);
+            PyObject* alpha_opA_dot_opB = alt_op_without_copy_%(float_type)s(0, (%(float_type)s*)PyArray_DATA((PyArrayObject*)opB_trans_dot_opA_trans), *M, *N, *M);
```
Comment (lamblin) [inline theano/tensor/alt_gemm_template.c]: I think it would be better to use `PyArray_Transpose` rather than get the pointer through `PyArray_DATA` and then creating a new array with `PyArray_New`.

**Comment by lamblin** [comment]
Could you also add the test case above to [`test_blas.py:TestBlasStrides`](https://github.com/Theano/Theano/blob/master/theano/tensor/tests/test_blas.py#L1786)?

File: theano/tensor/alt_gemm_template.c (hunk 14 above)
Comment (notoraptor) [inline theano/tensor/alt_gemm_template.c]: Done !

**Comment by notoraptor** [comment]
New update ! PyArray_Transpose now used in gemm function, and new test added to test_blas.py/TestBlasStrides (test_gemm_non_contiguous).

All tests succeeds ! Recall: these are the tests executed:

```
theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestAbstractConvNoOptim
theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestBilinearUpsampling
theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestCorrConv2d
theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestCorrConv3d
theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestCpuConv2d
theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/nnet/tests/test_abstract_conv.py:TestCpuConv3d
theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/tests/test_blas_c.py
theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/tests/test_blas.py
theano-cache purge && THEANO_FLAGS=blas.ldflags= nosetests theano/tensor/tests/test_blas_scipy.py
theano-cache purge && THEANO_FLAGS=optdb.max_use_ratio=10,blas.ldflags= nosetests theano/tensor/nnet/tests/test_corr3d.py
theano-cache purge && THEANO_FLAGS=optdb.max_use_ratio=7,blas.ldflags= nosetests theano/tensor/nnet/tests/test_corr.py

```
